kbs/src/plugins/implementations/pkcs11.rs

```diff
@@ -0,0 +1,332 @@
+            bail!("Slot index out of range");
+        }
+
+        let mut session = pkcs11.open_rw_session(slots[slot_index])?;
+        session.login(UserType::User, Some(&AuthPin::new(config.pin.clone())))?;
+
+        // Generate a UUID to for the wrapping keypair.
+        let wrapkey_id = Uuid::new_v4();
+
+        // Create the HSM wrapping keypair.
+        Pkcs11Backend::wrap_key_new(&mut session, &wrapkey_id)?;
+
+        Ok(Self {
+            session: Arc::new(Mutex::new(session)),
+            wrapkey_id,
+        })
+    }
+}
+
+#[async_trait::async_trait]
+impl ClientPlugin for Pkcs11Backend {
+    async fn handle(
+        &self,
+        body: &[u8],
+        _query: &str,
+        path: &str,
+        method: &Method,
+    ) -> Result<Vec<u8>> {
+        let desc = path
+            .strip_prefix('/')
+            .context("accessed path is illegal, should start with '/'")?;
+
+        match desc {
+            "wrap-key" => self.wrap_key_handle(body, method).await,
+            _ => {
+                let (action, params) = desc.split_once('/').context("accessed path is invalid")?;
+                match action {
+                    "resource" => self.resource_handle(params, body, method).await,
+                    _ => bail!("invalid path"),
+                }
+            }
+        }
+    }
+
+    async fn validate_auth(
+        &self,
+        _body: &[u8],
+        _query: &str,
+        _path: &str,
+        method: &Method,
+    ) -> Result<bool> {
+        match *method {
+            Method::GET => Ok(false),
+            Method::POST => Ok(true),
+            _ => bail!("invalid method"),
+        }
+    }
+
+    async fn encrypted(
+        &self,
+        _body: &[u8],
+        _query: &str,
+        _path: &str,
+        _method: &Method,
+    ) -> Result<bool> {
+        Ok(true)
+    }
+}
+
+#[async_trait::async_trait]
+impl StorageBackend for Pkcs11Backend {
+    async fn read_secret_resource(&self, resource_desc: ResourceDesc) -> Result<Vec<u8>> {
+        let session = self.session.lock().await;
+
+        // find object with matching label
+        let attributes = vec![Attribute::Label(Vec::from(resource_desc.to_string()))];
+        let objects = session.find_objects(&attributes)?;
+
+        if objects.is_empty() {
+            bail!(
+                "Could not find object with label {}",
+                resource_desc.to_string()
+            );
+        }
+        let object = objects[0];
+
+        // check that object has a readable value attribute
+        let value_attribute = vec![AttributeType::Value];
+        let attribute_map = session.get_attribute_info_map(object, value_attribute.clone())?;
+        let Some(AttributeInfo::Available(_size)) = attribute_map.get(&AttributeType::Value) else {
+            bail!("Key does not have value attribute available.");
+        };
+
+        // get the value
+        let value = &session.get_attributes(object, &value_attribute)?[0];
```

Comment:
looks like we are directly indexing the result here `[0]`, this would cause panic if the returned vector is empty. Could we use an error handling like
```rust
let value = &session.get_attributes(object, &value_attribute)?.get(0).ok_or(anyhow!("empty attributes returned"))?;
```